### PR TITLE
[macos] build: add build-number and buid-name arguments

### LINKED ISSUE #54542: Support --build-name and --build-number in build macOS
The values will be injected into the Info.plist as CFBundleShortVersionString and CFBundleVersion, respectively

https://github.com/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/macos/Runner/Info.plist#L19

### PR DESCRIPTION
**Pull request by hpoul**
## Description

This simply adds the options `--build-number` and `--build-name` to the `flutter build macos` command. It already supports writing the `buildInfo` into the generated files used by the macos app templates, so just enabling the options already passes it through to the macos build.

## Tests

I did not add any tests, please advise if tests are necessary and how to test this. I've tried looking at the build tests for iOS but couldn't find how the parameters are tested there?

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
@hpoul this change will need tests, for example see https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart

To start with, you should test that a command that passed in build-number and build-name leads to those being written to the right location.

Even though these arguments get written to the generated xcconfig, does anything in the macOS template use them?

**Comment by hpoul** [comment]
> Even though these arguments get written to the generated xcconfig, does anything in the macOS template use them?

@jonahwilliams thanks for taking a look, yes the plist file references the variables defined in the xcconfig: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/app/macos.tmpl/Runner/Info.plist#L19-L22

I am still not sure how to create those tests, I was not able to find how those parameters were tested on the iOS side, which implements them. I guess the best path would be to make sure that those values got written from the parameters into the generated xcconfig?

To be honest i'm pretty confused by those tests.. for example https://github.com/flutter/flutter/blob/74a1b9b33ed2c76f9808eeb70feec17aefedff97/packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart#L148-L162 "macOS build invokes xcode build (debug)" - I understand how the mock environment is set up and executed, but I can't figure out where the actual check happens that the debug flag is passed into the build command?

**Comment by jonahwilliams** [comment]
>  but I can't figure out where the actual check happens that the debug flag is passed into the build command?

Essentially the process manager will throw if there is a different command from what is expected.

At any rate, the following file will write the generated xcconfig file: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/macos/build_macos.dart#L36

So you just need to run the command in the fake environment, and then assert that this file contains the build/name build identifiers key.

Something like:

```
  testUsingContext('macOS build supports build name and number', () async {
    final BuildCommand command = BuildCommand();
    createMinimalMockProjectFiles();

    await createTestCommandRunner(command).run(
      const <String>['build', 'macos', '--debug', '--build-name='???, '--build-number=???']
    );
    final String contents = fileSystem.file('path/to/xcconfig').readAsStringSync();

    expect(contents, contains( /* build name and number */);
  }, overrides: <Type, Generator>{
    FileSystem: () => fileSystem,
    ProcessManager: () => FakeProcessManager.list(<FakeCommand>[
      setUpMockXcodeBuildHandler('Debug')
    ]),
    Platform: () => macosPlatform,
    FeatureFlags: () => TestFeatureFlags(isMacOSEnabled: true),
  });
```

**Comment by hpoul** [comment]
@jonahwilliams wow, thanks for taking the time to explain it.. that was much easier than I expected it.. I've now basically added the test exactly like you said.. and everything seems to work 👍️let's see if it also works in the automated test runs :)

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart [hunk 1]
```diff
@@ -191,6 +191,33 @@ void main() {
     Platform: () => macosPlatform,
     FeatureFlags: () => TestFeatureFlags(isMacOSEnabled: true),
   });
+  testUsingContext('macOS build supports build name and number', () async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:194]: nit: newline between previous test case
```suggestion

  testUsingContext('macOS build supports build name and number', () async {
```

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart [hunk 2]
```diff
@@ -191,6 +191,33 @@ void main() {
...
+      const <String>[
+        'build',
+        'macos',
+        '--debug',
+        '--build-name=1.2.3',
+        '--build-number=42',
+      ],
+    );
+    final String contents = fileSystem.file(
+        './macos/Flutter/ephemeral/Flutter-Generated.xcconfig')
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:208]: nit: indent 2 spaces and put `.file` on the same line as the path

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart (hunk 1 above)
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:194]:
```suggestion
  testUsingContext('macOS build supports build-name and build-number', () async {
```

File: packages/flutter_tools/lib/src/commands/build_macos.dart [hunk 3]
```diff
@@ -23,8 +23,9 @@ class BuildMacosCommand extends BuildSubCommand {
     addBuildModeFlags();
     addDartObfuscationOption();
     usesExtraFrontendOptions();
+    usesBuildNumberOption();
+    usesBuildNameOption();
   }
-
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/build_macos.dart:27]: nit: add this newline back

**Review by jonahwilliams — APPROVED** [review]
LGTM!

Thank you very much for the PR. The review comments I added are fairly minor

**Comment by jonahwilliams** [comment]
I also filled https://github.com/flutter/flutter/issues/54542 which we can close with this PR

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart (hunk 1 above)
Comment (hpoul) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart]: 👍️

Comment (hpoul) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart]: 👍️

File: packages/flutter_tools/lib/src/commands/build_macos.dart (hunk 3 above)
Comment (hpoul) [reply packages/flutter_tools/lib/src/commands/build_macos.dart:27]: 👍️

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart (hunk 2 above)
Comment (hpoul) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart]: 👍️

**Comment by hpoul** [comment]
thanks,  i've committed the requested changes..

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart [hunk 4]
```diff
@@ -192,6 +192,34 @@ void main() {
     FeatureFlags: () => TestFeatureFlags(isMacOSEnabled: true),
   });
 
+  testUsingContext('macOS build supports build-name and buid-number', () async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:195]:
```suggestion
  testUsingContext('macOS build supports build-name and build-number', () async {
```

**Review by jonahwilliams — COMMENTED** [review]
Small typo, otherwise great!

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart (hunk 4 above)
Comment (hpoul) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart]: 🤦️ force-pushed the change

**Comment by jonahwilliams** [comment]
@hpoul if all of the tests pass this will get automatically merged the next time the tree is green. If not, we'll circle back and I'll help figure out what failed

**Pull request merged**